pkg/collector/distinct_value_collector.go

```diff
@@ -52,28 +67,39 @@ func (d *DistinctValue[T]) Collect(v T) (exceeded bool) {
 	if d.limExceeded {
 		return true
 	}
-
 	// Calculate length
 	valueLen := d.len(v)
+	if d.maxDataSize > 0 && d.currDataSize+valueLen >= d.maxDataSize {
+		level.Warn(d.logger).Log("msg", "Max data exceeded", "dataSize", d.currDataSize, "maxDataSize", d.maxDataSize)
```

Comment:
i'm not sure logging lines at such a low level is useful. it would be difficult to pair this up with the related request to understand what happened.

if we want to log the reason for stopping early we should just store the reason as a string in the collector and then log it somewhere like here:

https://github.com/grafana/tempo/pull/4320/files#diff-a1fec687cd87c7167aee355f831ebae6beb368066e3bfbf092ef0aaf5e216aa2R568-R570

Reply:
i'm also disliking the amount of boilerplate necessary to pipe the logger object all the way down here. let's store the reason for ending early as suggested and then we can log it at a higher level.

Reply:
It is true that injecting the logger requires a massive amount of changes. I just wanted to propagate the context (the org_id).

Said so, I don't think this is in the hot path; these logs will only be written if we hit a stop condition. Therefore, it is irrelevant to the level they are. The main reason for keeping them there was to not leak the internal implementation of the collector. 

I believe this info can be handy to fine-tune and understand how the searches are working, also it give us useful data. I will store it in string.